lib/streamlit/elements/markdown.py

```diff
@@ -271,6 +277,85 @@ def divider(self) -> DeltaGenerator:
         divider_proto.element_type = MarkdownProto.Type.DIVIDER
         return self.dg._enqueue("markdown", divider_proto)
 
+    @gather_metrics("badge")
+    def badge(
+        self,
+        label: str,
+        *,  # keyword-only arguments:
+        icon: str | None = None,
+        color: Literal[
+            "blue",
+            "green",
+            "orange",
+            "red",
+            "violet",
+            "gray",
+            "grey",
+            "rainbow",
+            "primary",
+        ] = "blue",
+    ) -> DeltaGenerator:
+        """Display a colored badge with an icon and label.
+
+        You can also insert badges directly in Markdown, e.g. via
+        `st.markdown(":blue-badge[Home]")`. This works in all places where Streamlit
+        supports Markdown, e.g. widget labels or `st.table` cells. See `st.markdown`
+        for more information.
+
+        Parameters
+        ----------
+        label : str
+            The text to display in the badge.
+
+        icon : str or None
+            An optional emoji or icon to display next to the badge label. If
+            ``icon`` is ``None`` (default), no icon is displayed. If ``icon``
+            is a string, the following options are valid:
+
+            - A single-character emoji. For example, you can set ``icon="🚨"``
+              or ``icon="🔥"``. Emoji short codes are not supported.
+
+            - An icon from the Material Symbols library (rounded style) in the
+              format ``":material/icon_name:"`` where "icon_name" is the name
+              of the icon in snake case.
+
+              For example, ``icon=":material/thumb_up:"`` will display the
+              Thumb Up icon. Find additional icons in the `Material Symbols \
+              <https://fonts.google.com/icons?icon.set=Material+Symbols&icon.style=Rounded>`_
+              font library.
+
+        color : str
+            The color to use for the badge. Supported colors are: blue, green,
+            orange, red, violet, gray/grey, primary.
+            If you use "primary" for color, Streamlit will use the default
+            primary accent color unless you set the ``theme.primaryColor``
+            configuration option.
+
+        Examples
+        --------
+        >>> import streamlit as st
+        >>>
+        >>> # Simple badge
+        >>> st.badge("Home")
+        >>>
+        >>> # Badge with icon and color
+        >>> st.badge("Success", icon=":material/check:", color="green")
+        >>>
+        >>> # Multiple badges side by side in Markdown
+        >>> st.markdown(
+        ...     "Here are some badges: :orange-badge[⭐️ Favorite] :blue-badge[🏠 Home] :green-badge[✅ Success]"
+        ... )
+        """
+        if icon is not None:
+            icon_str = validate_icon_or_emoji(icon) + " "
+        else:
+            icon_str = ""
+
+        badge_proto = MarkdownProto()
+        badge_proto.body = f":{color}-badge[{icon_str}{label}]"
```

Comment:
I think we might need to remove all `[` and `]` characters from `label` so that it doesn't unexpectedly breaks.

Reply:
Great point, I escaped them.